### ISSUE TITLE
[BUG] SIGHUP causes deadlocks

### ISSUE THREAD

**Issue by zzp-08** [rspamd/rspamd#2959]
When reload rspamd using kill -s HUP main_rspamd_id, there is a very small probability of deadlock.
At this point, the owner of the lock is an exiting zombie process (rspamd main has not yet recovered resources).
The main rspamd process is blocked in the callback handling SIGHUP.
The main rspamd process status is:
0x00007f066e622f8c in __lll_robust_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
(gdb) bt
#0  0x00007f066e622f8c in __lll_robust_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f066e61e373 in _L_robust_lock_202 () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x00007f066e61dcff in __pthread_mutex_lock_full () from /lib/x86_64-linux-gnu/libpthread.so.0
#3  0x000000000046af90 in ?? ()
#4  0x000000000046bf17 in ?? ()
#5  0x000000000046cc41 in rspamd_common_logv ()
#6  0x000000000046ce4d in rspamd_default_logv ()
#7  0x000000000046ced8 in rspamd_default_log_function ()
#8  0x00007f066df36d00 in g_hash_table_foreach () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#9  0x00000000004432af in ?? ()
#10 0x00007f066da887e5 in event_base_loop () from /usr/lib/x86_64-linux-gnu/libevent-2.0.so.5
#11 0x0000000000434c64 in main ()
(gdb) 

The normal rspamd process status is:
0x00007f7664885f8c in __lll_robust_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
(gdb) bt
#0  0x00007f7664885f8c in __lll_robust_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f7664881373 in _L_robust_lock_202 () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x00007f7664880cff in __pthread_mutex_lock_full () from /lib/x86_64-linux-gnu/libpthread.so.0
#3  0x000000000046af90 in ?? ()
#4  0x000000000046bf17 in ?? ()
#5  0x000000000046cc41 in rspamd_common_logv ()
#6  0x000000000046ce4d in rspamd_default_logv ()
#7  0x000000000046ced8 in rspamd_default_log_function ()
#8  0x00000000004f9977 in ?? ()
#9  0x00000000004f9b68 in rspamd_redis_pool_connect ()
#10 0x00000000004a16f0 in ?? ()
#11 0x00000000004a1bc4 in ?? ()
#12 0x00000000005b1a66 in ?? ()
#13 0x00000000005a16e0 in lua_pcall ()
#14 0x00000000004a2bee in ?? ()
#15 0x000000000056d946 in redisProcessCallbacks ()
#16 0x00007f7663ceb254 in event_base_loop () from /usr/lib/x86_64-linux-gnu/libevent-2.0.so.5
#17 0x00000000004445c7 in start_worker ()
#18 0x00000000004f5860 in rspamd_fork_worker ()
#19 0x0000000000441c43 in ?? ()
#20 0x0000000000442de4 in ?? ()
#21 0x00007f7663ceb7e5 in event_base_loop () from /usr/lib/x86_64-linux-gnu/libevent-2.0.so.5
#22 0x0000000000434c64 in main ()
(gdb) 

### Versions
1.8.2  Debian

**Comment by vstakhov** [comment]
Just bear in mind that 1.8.2 version is not supported. However, it might be still affecting the actual version (and/or master branch). I will check.

**Comment by zzp-08** [comment]
Should rspamd_hup_handler be added to rspamd_log_nolock? I'm verifying.

**Comment by polarathene** [comment]
**TL;DR:** Probably safe to ignore this comment. I'm not encountering the issue myself but made note of this a few years back and as there doesn't appear to be any confirmation of a follow-up since I've added some context below for reference.

---

> When reload rspamd using `kill -s HUP main_rspamd_id`, there is a very small probability of deadlock.
> - At this point, the owner of the lock is an exiting zombie process (rspamd main has not yet recovered resources).
> - The main rspamd process is blocked in the callback handling SIGHUP.

@vstakhov , just curious if you ever looked into this ~~and if `SIGHUP` is valid or should be avoided~~? (_**EDIT:** Seems to be valid and encouraged, unclear if this deadlock issue was resolved_)

- [None of the code specifically referring to `SIGHUP` has been touched since](https://github.com/search?q=repo%3Arspamd%2Frspamd+SIGHUP&type=code) the discussion here in 2019.
- [Nor any new commits](https://github.com/search?q=repo%3Arspamd%2Frspamd+SIGHUP&type=commits) (_last one matching query was Aug 2016 which has an associated [changelog fix entry for `SIGHUP`](https://github.com/rspamd/rspamd/blob/eed249e7315e47ccbb5a33b3e3adab120b9f60cb/ChangeLog#L4769) with the Rspamd `1.3.4` release_).
- Bug Reports:
  - [Aug 2021 issue referenced an problem with `SIGHUP`](https://github.com/rspamd/rspamd/issues/3853#issuecomment-908182153) which was [resolved](https://github.com/rspamd/rspamd/issues/3853#issuecomment-908187852) (issue with a pointer). Unclear if this was related to `SIGHUP`.
  - [Dec 2020 issue comment encouraging reload via `SIGHUP`](https://github.com/rspamd/rspamd/issues/3562#issuecomment-741636730).
  - 2019 - This issue.
  - [Aug 2017 `SIGHUP` racey failure](https://github.com/rspamd/rspamd/issues/1758#issuecomment-321492956).
  - [Apr 2017 `SIGHUP` encouraged as in better shape](https://github.com/rspamd/rspamd/issues/825#issuecomment-291122640), in that [same issue (but an Aug 2016 comment)](https://github.com/rspamd/rspamd/issues/825#issuecomment-238490335) it was acknowledged that `SIGHUP` support was broken.
